File: CentralServer/services/moviesService.js
const axios = require('axios');

const SPRING_SERVICE_URL='http://localhost:8080';
const MONGO_SERVICE_URL='http://localhost:3001';

/**
 * Retrieves a list of movie base on provided options
 *
 * @async
 * @function getMovies
 * @param {Object} opt query options for filtering movies
 * @param {string} opt.sort sorting field (e.g., 'popular', 'rating')
 * @param {string} opt.order sorting order ('ASC' or 'DSC')
 * @param {number} opt.page page number for pagination (zero-based)
 * @param {number} opt.size number of results per page
 * @param {string[]} opt.genres list of genres to filter by
 * @return {Promise<Object>} list of movies
 * @throws {Error} if fetching movies fails
 */
exports.getMovies = async (opt) => {
    let {sort, order, page, size, genres, genre, year} = opt;

    try {
        const response = await axios.get(SPRING_SERVICE_URL + "/movies/options", {
            params: {
                sort,
                order,
                page,
                size,
                genres: genres || genre,
                year
            }
        });
        return response.data;
    } catch (err) {
        console.error("Error during movies request:", err.message);
        throw new Error("Failed to fetch movies from server");
    }
}

/**
 * Retrieves a paginated list of movies matching the specified title.
 *
 * @async
 * @function getTitle
 * @param {Object} data query parameters
 * @param {string} data.filmTitle the title of the movie to search for
 * @param {number} data.page the page number to retrieve (zero-based)
 * @param {number} data.size the number of results per page
 * @return {Promise<Object>} paginated list of movies
 * @throws {Error} if fetching movies fails
 */
exports.getTitle = async (data) => {
    let {filmTitle, page, size} = data;
    try {
        const path = `/movies/name/${encodeURIComponent(filmTitle)}`;
        const response = await axios.get(SPRING_SERVICE_URL + path, {params:{page, size}});

        return response.data;
    } catch (err) {
        console.error("Error during movies title request:" + err.message);
        throw new Error("Failed to fetch movies title from server");
    }
}

/**
 * Retrieves a paginated list of movies belonging to a specific genre.
 *
 * @async
 * @function getGenre
 * @param {Object} data query parameters
 * @param {string} data.genre the genre to filter movies by
 * @param {number} data.page the page number to retrieve (zero-based)
 * @param {number} data.size the number of results per page
 * @return {Promise<Object>} paginated list of movies
 * @throws {Error} if fetching movies fails
 */
exports.getGenre = async (data) => {
    let {genre, page, size} = data;
    try {
        const path = `/movies/genre/${encodeURIComponent(genre)}`;
        const response = await axios.get(SPRING_SERVICE_URL + path, {params:{page, size}});

        return response.data;
    } catch (err) {
        console.error("Error during movies genre request:" + err.message);
        throw new Error("Failed to fetch movies genre from server");
    }
}

/**
 * Retrieves a paginated list of movies released in a specific year.
 *
 * @async
 * @function getYear
 * @param {Object} data query parameters
 * @param {string} data.year the year of release to filter movies
 * @param {number} data.page the page number to retrieve (zero-based)
 * @param {number} data.size the number of results per page
 * @return {Promise<Object>} paginated list of movies
 * @throws {Error} if fetching movies fails
 */
exports.getYear = async (data) => {
    let {year, page, size} = data;
    try {
        const path = `/movies/year/${year}`;
        const response = await axios.get(SPRING_SERVICE_URL + path, {params:{page, size}});

        return response.data;
    } catch (err) {
        console.error("Error during movies year request:" + err.message);
        throw new Error("Failed to fetch movies by year from server");
    }
}

/**
 * Retrieves a list of related movies associated with a specific movie ID based on shared themes.
 *
 * @async
 * @function getRelatedMovies
 * @param {Object} data query parameters
 * @param {string} data.movieId the unique identifier of the movie to find related films for
 * @param {number} data.page the page number to retrieve (zero-based)
 * @param {number} data.size the number of results per page
 * @return {Promise<Object>} paginated list of related movies
 * @throws {Error} if fetching related movies fails
 */
exports.getRelatedMovies = async (data) => {
    let {movieId, page, size} = data;
    try {
        const response = await axios.get(SPRING_SERVICE_URL + "/movies/related", {params:{movieId, page, size}});

        return response.data;
    } catch (err) {
        console.error("Error during related movies request:" + err.message);
        throw new Error("Failed to fetch related movies from server");
    }
}

/**
 * Retrieves a paginated list of movies with the specified rating.
 *
 * @async
 * @function getRating
 * @param {Object} data query parameters
 * @param {number} data.rating the rating to filter movies
 * @param {number} data.page the page number
 * @param {number} data.size the number of results per page
 * @return {Promise<Object>} list of movies
 * @throws {Error} if fetching movies fails
 */
exports.getRating = async (data) => {
    let {rating, page, size} = data;
    try {
        const path = `/movies/rating/${rating}`;
        const response = await axios.get(SPRING_SERVICE_URL + path, {params:{page, size}});

        return response.data;
    } catch (err) {
        console.error("Error during movies rating request:", err.message);
        throw new Error("Failed to fetch movies by rating from server");
    }
}

/**
 * Retrieves a list of movies featuring a specific Oscar-winning actor.
 *
 * @async
 * @function getOscarActor
 * @param {Object} data query parameters
 * @param {string} data.actor the name of the Oscar-winning actor
 * @return {Promise<Object>} list of movies
 * @throws {Error} if fetching movies fails
 */
exports.getOscarActor = async (data) => {
    let {actor} = data;
    try {
        const path = `/movies/actor/${encodeURIComponent(actor)}`;
        const response = await axios.get(SPRING_SERVICE_URL + path);

        return response.data;
    } catch (err) {
        console.error("Error during Oscar actor movies request:", err.message);
        throw new Error("Failed to fetch movies by Oscar winner from server");
    }
}

File: SpringBootServer/src/main/java/it/unito/iumtweb/springboot/movies/MoviesService.java
package it.unito.iumtweb.springboot.movies;

import it.unito.iumtweb.springboot.themes.Themes;
import it.unito.iumtweb.springboot.themes.ThemesRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 *  Service class that provides high-level operations for retrieving and analyzing
 * Movies data. This class acts as an intermediate layer between the controller
 * and the repository, converting entities into DTOs when needed.
 */
@Service
public class MoviesService {
    private final MoviesRepository moviesRepo;
    private final ThemesRepository themesRepo;

    /**
     * Constructs a new MoviesService with the specified MoviesRepository and ThemesRepository.
     *
     * @param moviesRepo the repository for movie entities
     * @param themesRepo the repository for theme entities
     */
    public MoviesService(MoviesRepository moviesRepo, ThemesRepository themesRepo) {
        this.moviesRepo = moviesRepo;
        this.themesRepo = themesRepo;
    }

    /**
     * Retrieves a paginated list of movies featuring the specified actor, ignoring case.
     * The results are paginated and sorted by rating in descending order.
     *
     * @param name the name of the actor (case-insensitive)
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a list of {@link MoviesCardInfoDTO} objects for movies featuring the actor
     */
    public Page<MoviesCardInfoDTO> getMovieByActor(String name, int page, int size) {
        Pageable pageable = PageRequest.of(page,size, Sort.by(Sort.Direction.DESC,"rating"));
        return moviesRepo.findMoviesByActor(name, pageable);
    }

    /**
     * Retrieves a paginated list of movies whose titles contain the specified string, ignoring case.
     * The results are paginated and sorted by rating in descending order.
     *
     * @param title the title or partial title to search for (case-insensitive)
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a list of {@link MoviesCardInfoDTO} objects matching the search criteria
     */
    public Page<MoviesCardInfoDTO> getMovieByTitle(String title,int page, int size) {
        Pageable pageable = PageRequest.of(page,size,Sort.by(Sort.Direction.DESC,"rating"));
        return moviesRepo.findMoviesByTitle(title, pageable);
    }

    /**
     * Retrieves a paginated list of movies filtered by the specified genre.
     * The results are paginated and sorted by rating in descending order.
     *
     * @param genre the genre to filter movies
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a list of {@link MoviesCardInfoDTO} objects for movies filtered by genre
     */
    Page<MoviesCardInfoDTO> getMoviesByGenre(String genre, int page, int size) {
        Pageable pageable = PageRequest.of(page,size,Sort.by(Sort.Direction.DESC,"rating"));
        return moviesRepo.findMoviesByGenre(genre, pageable);
    }

    /**
     * Retrieves a paginated list of movies released in the specified year.
     * The results are paginated and sorted by movie name in descending order.
     *
     * @param year the year of release to filter movies
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a list of {@link MoviesCardInfoDTO} objects for movies released in the specified year
     */
    public Page<MoviesCardInfoDTO> getMoviesByYear(String year, Integer page, Integer size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("name").descending());
        return moviesRepo.findMoviesByDate(year, pageable);
    }

    /**
     * Retrieves a paginated list of movies with the specified rating.
     * The results are paginated and sorted by rating in descending order.
     *
     * @param rating the rating to filter movies
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a list of {@link MoviesCardInfoDTO} objects for movies with the specified rating
     */
    public Page<MoviesCardInfoDTO> getMoviesByRating(Double rating, Integer page, Integer size) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("rating").descending());
        return moviesRepo.findMoviesByRating(rating, pageable);
    }

    /**
     * Retrieves a paginated list of movies filtered and sorted according to the provided options.
     * Allows sorting by rating, year of release, or movie title, and filtering by genres.
     *
     * @param sort the sorting criteria ("newest", "movie", or default "rating")
     * @param genres a list of genres to filter by; if null or empty, no genre filtering is applied
     * @param order the sort order ("ASC" for ascending, otherwise descending)
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a {@link Page} of {@link MoviesCardInfoDTO} objects matching the criteria
     */
    public Page<MoviesCardInfoDTO> getMoviesByOptions(String sort, String genres, String year, String order, Integer page, Integer size) {


        String sorting = switch (sort) {
            case "year", "newest" -> "date";
            case "movie", "title" -> "name";
            default -> "rating";
        };

        Sort.Direction direction = (order != null && order.equalsIgnoreCase("ASC")) ? Sort.Direction.ASC : Sort.Direction.DESC;

        
        Pageable pageable = PageRequest.of(page, size, Sort.by(direction, sorting).and(Sort.by(Sort.Direction.ASC, "movie_id")));


        String genreParam = (genres == null || genres.isEmpty() ||  genres.equals("all"))
                ? null : genres;


        String yearParam = (year == null || year.trim().isEmpty() || year.equalsIgnoreCase("all"))
                ? null : year;


        return moviesRepo.findMoviesWithSmartFilters(genreParam, yearParam, pageable);
    }

    List<MoviesCardInfoDTO> getMoviesByActorOscarWinner(String name) {
        return moviesRepo.findMoviesByActorOscarWinner(name);
    }

    /**
     * Finds movies similar to the specified movie based on shared themes.
     * Retrieves a paginated list of movies that share at least 5 themes with the given movie,
     * excluding the movie itself.
     *
     * @param movieId the unique identifier of the reference movie
     * @param page the page number to retrieve (zero-based)
     * @param size the number of results per page
     * @return a list of {@link MoviesCardInfoDTO} objects representing related movies
     */
    public List<MoviesCardInfoDTO> getRelatedMoviesByThemes(String movieId, Integer page, Integer size) {
        List<Themes> themes = themesRepo.findThemesByMovieId(movieId);
        List<String> themesNames = themes.stream().map(Themes::getTheme).toList();

        Pageable pageable = PageRequest.of(page, size);
        Page<MoviesCardInfoDTO> pageOfMovies = moviesRepo.findMoviesWithMinimumThemes(themesNames, movieId, 5, pageable);

        System.out.println("Related movies: " + pageOfMovies.getContent());
        return pageOfMovies.getContent();
    }


    /**
     * shows movies details on click
     * @param movieId
     * @return
     */
    public Page<MoviesCardInfoDTO> getMoviesDetail(String movieId) {
        // Usa moviesRepo (istanza)
        Pageable pageable = PageRequest.of(0, 1);
        Page<MoviesCardInfoDTO> result = moviesRepo.findMovieDetailById(movieId,pageable);
        return result;
    }

    /**
     * Adds a new movie to the database.
     *
     * @param movie the {@link Movies} entity to be saved
     */
    public void addMovie(Movies movie) {
        moviesRepo.save(movie);
    }

    /**
     * Retrieves all movies from the database.
     *
     * @return a list of all {@link Movies} entities
     */
    public List<Movies> getAllMovies() {
        return moviesRepo.findAll();
    }



    public List<String> getAllYears() {
        return moviesRepo.findAllDistinctYears();
    }


}
